File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode parent = new ListNode(-1, head);
        int length = 0;
        ListNode temp = head;
        if(head == null) return head;
        while(temp != null) {
            length++;
            temp = temp.next;
        }
        System.out.println(length);
        temp = parent;
        for(int i = 0; i < length - n; i++) {
            temp = temp.next;
        }

        temp.next = temp.next.next;
        return parent.next;
    }
}

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        random = {}
        old_to_new = {}

        temp = head
        copy = Node(-1)
        iter = copy
        while temp is not None:
            iter.next = Node(temp.val)
            iter = iter.next
            old_to_new[temp] = iter
            temp = temp.next
        
        temp = head
        iter = copy.next
        while iter != None:
            if temp.random is not None:
                iter.random = old_to_new[temp.random]
            else:
                iter.random = None
            iter = iter.next
            temp = temp.next
        
        return copy.next

        
        